tools/clang/lib/CodeGen/BackendUtil.cpp

```diff
@@ -762,7 +763,18 @@ void clang::EmitBackendOutput(DiagnosticsEngine &Diags,
                               raw_pwrite_stream *OS) {
   EmitAssemblyHelper AsmHelper(Diags, CGOpts, TOpts, LOpts, M);
 
-  AsmHelper.EmitAssembly(Action, OS);
+  // Catch any fatal errors during optimization passes here
+  // so that future passes can be skipped.
+  try {
+    AsmHelper.EmitAssembly(Action, OS);
+  } catch (const ::hlsl::Exception &hlslException) {
+    const char *msg = hlslException.what();
+    const std::string msgStr(msg);
+    unsigned DiagID = Diags.getCustomDiagID(
+        DiagnosticsEngine::Error,
+        "'%0'\nFatal error during optimization, aborting.");
```

Comment:
I don't think surrounding the msgStr in quotes `'` is necessary.
Seeing this in the example, I'm thinking that we probably don't really need the extra text here either.
Since you can create single-use things on the fly, I don't see why we couldn't just reduce the code in the block to:
```
Diags.Report(Diags.getCustomDiagID(DiagnosticsEngine::Error, "%0")) << StringRef(hlslException.what());
```